yarn-project/acir-simulator/src/client/simulator.ts

```diff
@@ -20,12 +22,31 @@ import { UnconstrainedFunctionExecution } from './unconstrained_execution.js';
  * The ACIR simulator.
  */
 export class AcirSimulator {
+  private static solver: WasmBlackBoxFunctionSolver; // ACVM's backend
   private log: DebugLogger;
 
   constructor(private db: DBOracle) {
     this.log = createDebugLogger('aztec:simulator');
   }
 
+  /**
+   * Gets or initializes the ACVM WasmBlackBoxFunctionSolver.
+   *
+   * @remarks
+   *
+   * Occurs only once across all instances of AcirSimulator.
+   * Speeds up execution by only performing setup tasks (like pedersen
+   * generator initialization) one time.
+   * WARNING: it is unclear whether this will work in a multi-threaded
```

Comment:
```
   * WARNING: it is unclear whether this will work in a multi-threaded
   * environment where multiple threads seek to use the same backend.
```
Technically not an issue js can have, but we can have async functions interleave improperly. This is why some stuff uses explicit sequencing like our merkle tree operation list

Reply:
This might be best as a TODO

Reply:
The TODO likely should just have a full lock on this. So a lock that you get when you start, release when done (in a finally block, ideally) and anyone who tries to grab the lock to start a series of ops would wait

Reply:
Actually, the ideal is that the functions are stateless other than constant lookup data. Then it shouldn't matter

Reply:
If we have internal *global* state then it matters. otherwise it should be fine!

Reply:
This is why e.g. we have no problems with our circuits singleton as all functions are pure input -> output

Reply:
To be honest, I am not even sure that this actually is an issue? Maybe we _can_ use this solver in simultaneous calls to the ACVM.... not sure?

Reply:
Created #1627